src/main/java/org/interledger/ilp/exceptions/InterledgerException.java

```diff
@@ -1,36 +1,196 @@
+     * can be used to create an updated forwardedBy list right before calling the constructor.
+     * 
+     * @param errCode,    one of the defined InterledgerException.ErrorCode defined.
+     * @param triggeredBy ILP address (prefix if it's a ledger) of the entity that originally emitted the error
+     * @param forwardedBy ILP address list of connectors that relayed the error message
+     * @param triggeredAt Time when the error was initially emitted
+     * @param data        OCTET STRINGSIZE(0..8192) Unless otherwise specified, data SHOULD be encoded as UTF-8. 
+     *                    Protocols built on top of ILP SHOULD specify the encoding format of error data 
+     */
+    public InterledgerException(
+        ErrorCode errCode,
+        InterledgerAddress triggeredBy,
+        InterledgerAddress[] forwardedBy,
+        ZonedDateTime triggeredAt,
+        String data) {
+        super();
+        this.errCode     = Objects.requireNonNull(errCode    , "errCode     must not be null");
+        this.triggeredBy = Objects.requireNonNull(triggeredBy, "triggeredBy must not be null");
+        this.forwardedBy = Objects.requireNonNull(forwardedBy, "forwardedBy must not be null");
+        this.triggeredAt = Objects.requireNonNull(triggeredAt, "triggeredAt must not be null");
+        this.data        = Objects.requireNonNull(data       , "data        must not be null");
+    }
+
+    /**
+     * Constructs an instance of <code>InterledgerException</code>
+     * with default parameters for forwardedBy (Empty list) and 
+     * triggeredAt (ZonedDateTime.now()). 
+     *  In most situations such values match the default ones 
+     * when triggering a new exception (vs an exception received
+     * from another ILP node that is being forwarded back to 
+     * originating request clients)
+     * 
+     * Check the RFC https://interledger.org/rfcs/0003-interledger-protocol/#errors
+     * for the newest updated doc.
+     */
+    public InterledgerException(
+        ErrorCode errCode,
+        InterledgerAddress triggeredBy,
+        String data) {
+        this(errCode, triggeredBy, new InterledgerAddress[]{}, ZonedDateTime.now(), data);
+    }
+
+    public ErrorCode getErrCode() {
+        return errCode;
+    }
+
+    /**
+     * @return ErrorType categorizes the error as FINAL, TEMPORARY or RELATIVE
+     *         (See RFC for more info)
+     */
+    public ErrorType getErrorType(){
+        return errCode.errorType;
+    }
+
+    public InterledgerAddress getTriggeredBy() {
+        return triggeredBy;
+    }
+
+    public InterledgerAddress[] getForwardedBy() {
+        return forwardedBy;
+    }
+
+    public ZonedDateTime getTriggeredAt() {
+        return triggeredAt;
+    }
+
+    public String getData() {
+        return data;
+    }
+
+    // Utility static methods 
+
+    public static InterledgerAddress[] addSelfToForwardedBy(InterledgerAddress[] forwardedBy, InterledgerAddress selfAddress) {
```

Comment:
@sappenin This is a old debate. 
Must we use interfaces for *everything* and then implement them in concrete clases?

My point of view is that interfaces must be used to describe behaviour whose implementation can vary for different reasons (security, performance, memory, ...) while concrete classes must be used to describe data-centric (DTO) objects. Class instances could be interpreted as "rows" in a database table. (In the case of exception to lines of logs in syslog maybe). If the RFC mandates an object that must have some well defined fields (errCode that is 3 chars long followed by an InterledgerAddress ...) then defining an interface will not help at all. It will promote "random" implementations of a concrete schema that was well defined from the scratch.  Different implementation are not expected to improve the code. A final developer will expect to have a "box" (class) where he can place the data. This is similar to the java.security.spec classes. Is not expected for Java Security implementators to reimplement those "data objects". (but it expected that many different ways to fill them exists).